src/main/java/uk/nhs/digital/nhsconnect/lab/results/translator/mapper/OrganizationMapper.java

```diff
@@ -1,20 +1,79 @@
 import org.springframework.stereotype.Component;
 import uk.nhs.digital.nhsconnect.lab.results.model.edifact.Message;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.PerformerNameAndAddress;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.segmentgroup.InvolvedParty;
+import uk.nhs.digital.nhsconnect.lab.results.utils.UUIDGenerator;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static uk.nhs.digital.nhsconnect.lab.results.model.edifact.ServiceProviderCode.DEPARTMENT;
+import static uk.nhs.digital.nhsconnect.lab.results.model.edifact.ServiceProviderCode.ORGANIZATION;
 
 @Component
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
 public class OrganizationMapper {
-    public Organization mapToPerformingOrganization(final Message message) {
-        return new Organization();
+
+    private final UUIDGenerator uuidGenerator;
+
+    public Optional<Organization> mapToRequestingOrganization(final Message message) {
+        return message.getInvolvedParties().stream()
+            .filter(party -> party.getServiceProvider().getServiceProviderCode().equals(ORGANIZATION))
+            .map(InvolvedParty::getRequesterNameAndAddress)
+            .flatMap(Optional::stream)
+            .findFirst()
+            .map(requester -> mapToOrganization(requester.getOrganizationName(), null));
+    }
+
+    public Optional<Organization> mapToPerformingOrganization(final Message message) {
+        Stream<InvolvedParty> organizationStream = message.getInvolvedParties().stream()
+            .filter(party -> party.getServiceProvider().getServiceProviderCode().equals(ORGANIZATION));
+
+        String performingOrganizationName = mapToPerformerName(organizationStream);
+
+        if (performingOrganizationName != null) {
+            Stream<InvolvedParty> departmentStream = message.getInvolvedParties().stream()
+                .filter(party -> party.getServiceProvider().getServiceProviderCode().equals(DEPARTMENT));
+
+            String performingOrganizationDepartmentName = mapToPerformerName(departmentStream);
+
+            return Optional.of(mapToOrganization(performingOrganizationName, performingOrganizationDepartmentName));
+        }
+
+        return Optional.empty();
     }
 
-    public Organization mapToRequestingOrganization(final Message message) {
-        return new Organization();
+    private String mapToPerformerName(Stream<InvolvedParty> involvedPartyStream) {
+        return involvedPartyStream.map(InvolvedParty::getPerformerNameAndAddress)
+            .flatMap(Optional::stream)
+            .map(PerformerNameAndAddress::getOrganizationName)
+            .findFirst()
+            .orElse(null);
     }
 
-    public Organization mapToSpecimenCollectingOrganization(final Message message) {
-        return new Organization();
+    private Organization mapToOrganization(final String organizationName, final String departmentName) {
+        final var organization = new Organization();
```

Comment:
Can an organisation have an identifier or address? As i noticed some of the examples here: https://developer.nhs.uk/apis/itk3nationalpathology-1-1-0/design_singletest_bundle.html contain both of these. Also the fhir test data we created - some organisations contain identifiers.

Reply:
The spec states that the identifier is 
* required if it's a message recipient organisation
* advised if it's a practitioner who's a lab service requester, message recipient or a lab service provider 
* is not used when it's a lab service provider that's a department or a organisation.

Reply:
Thanks. So we're yet to map the message recipient organisation in here?

Reply:
In that [link](https://developer.nhs.uk/apis/itk3nationalpathology-1-1-0/design_singletest_bundle.html ), the organisation is a Requesting Organisation (`NAD+PO`) so yes it can have an identifier, but performing organisations (`NAD+SLA`) don't have an identifier

Reply:
Ok so we may need to add the optional identifier mapping?

Reply:
Oddly, the spec doesn't mention Requesting Organisations so I'm not sure if we need to add the optional identifier or not.

Reply:
Now I'm lost :D if we look at the spreadsheet, the identifier is mandatory for both performing and requesting organisations

Reply:
Oh, I was referring to the Word doc with the Edifact spec for the identifier rules for `NAD+PO` and `NAD+SLA`. 

Yeah, the spreadsheet says the identifier is mandatory

Reply:
I would go more with what the EDIFACT spec says (rather than the spreadsheet). EDIFACT spec should be our source of truth for now.
So in summary, the spec doesn't mention identifier for Requesting Organization (though it does in the Pathology model spreadsheet for Message Recipient Organization, which we ignore), and it says it's not required for a providing/performing organization. I'm thinking we leave this for now (so it doesn't block this story any further), and it can be addressed in a later story if it's a problem

Reply:
Ok, sounds like a plan 🙂